contracts/HATVault.sol

```diff
@@ -0,0 +1,816 @@
+        }
+    }
+
+    /* -------------------------------------------------------------------------------- */
+
+    /* --------------------------------- Helpers -------------------------------------- */
+
+    /**
+    * @dev Deposit funds to the vault. Can only be called if the committee had
+    * checked in and deposits are not paused.
+    * @param caller Caller of the action (msg.sender)
+    * @param receiver Reciever of the shares from the deposit
+    * @param assets Amount of vault's native token to deposit
+    * @param shares Respective amount of shares to be received
+    */
+    function _deposit(
+        address caller,
+        address receiver,
+        uint256 assets,
+        uint256 shares
+    ) internal override virtual nonReentrant {
+        if (!committeeCheckedIn)
+            revert CommitteeNotCheckedInYet();
+        if (shares == 0) revert AmountToDepositIsZero();
+        if (withdrawEnableStartTime[receiver] != 0 && receiver == caller) {
+            // clear withdraw request if caller deposits in her own account
+            withdrawEnableStartTime[receiver] = 0;
+        }
+
+        super._deposit(caller, receiver, assets, shares);
+    }
+
+    function _withdraw(
+        address caller,
+        address receiver,
+        address owner,
+        uint256 assets,
+        uint256 shares,
+        uint256 fee
+    ) internal nonReentrant {
+        if (assets == 0) revert WithdrawMustBeGreaterThanZero();
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        _burn(owner, shares);
+
+        IERC20 asset = IERC20(asset());
+        if (fee > 0) {
+            asset.safeTransfer(registry.owner(), fee);
+        }
+        asset.safeTransfer(receiver, assets);
+
+        emit Withdraw(caller, receiver, owner, assets, shares);
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 amount
+    ) internal virtual override {
+        if (to != address(0)) {
+            if (registry.isEmergencyPaused()) revert SystemInEmergencyPause();
+            // Cannot transfer or mint tokens to a user for which an active withdraw request exists
+            // because then we would need to reset their withdraw request
+            if (withdrawEnableStartTime[to] != 0) {
+                IHATVaultsRegistry.GeneralParameters memory generalParameters = registry.getGeneralParameters();
+                // solhint-disable-next-line not-rely-on-time
+                if (block.timestamp < withdrawEnableStartTime[to] + generalParameters.withdrawRequestEnablePeriod)
+                    revert CannotTransferToAnotherUserWithActiveWithdrawRequest();
+            }
+            rewardController.commitUserBalance(to, amount, true);
+        }
+
+        if (from != address(0)) {
+            if (activeClaim.createdAt != 0) revert ActiveClaimExists();
+            if (!_isWithdrawEnabledForUser(from)) revert InvalidWithdrawRequest();
```

Comment:
seems to me this is redundant, because this check is done on maxRedeem/maxWithdraw

Reply:
or we can remove this check from `withdraw` and `redeem`

Reply:
We need it in maxWithdraw and maxRedeem to comply with ERC4626 specs, and here to make sure it is checked also on token transfers.

Reply:
Or actually we could do something like this: https://github.com/hats-finance/hats-contracts/pull/347